redbot/core/utils/views.py

```diff
@@ -0,0 +1,258 @@
+            discord.SelectOption(label=_("Page {num}").format(num=num + 1), value=num)
+            for num, x in enumerate(pages)
+        ]
+        self.stop_button = _StopButton(
+            discord.ButtonStyle.red, "\N{HEAVY MULTIPLICATION X}\N{VARIATION SELECTOR-16}"
+        )
+        self.select_menu = self._get_select_menu()
+        self.add_item(self.stop_button)
+        if self.source.is_paginating() and not self.use_select_only:
+            self.add_item(self.first_button)
+            self.add_item(self.backward_button)
+            self.add_item(self.forward_button)
+            self.add_item(self.last_button)
+        if self.use_select_menu and self.source.is_paginating():
+            if self.use_select_only:
+                self.remove_item(self.stop_button)
+                self.add_item(self.select_menu)
+                self.add_item(self.stop_button)
+            else:
+                self.add_item(self.select_menu)
+
+    @property
+    def source(self):
+        return self._source
+
+    async def on_timeout(self):
+        if self.delete_after_timeout and not self.message.flags.ephemeral:
+            await self.message.delete()
+        elif self.disable_after_timeout:
+            for child in self.children:
+                child.disabled = True
+            await self.message.edit(view=self)
+        else:
+            await self.message.edit(view=None)
+
+    def _get_select_menu(self):
+        # handles modifying the select menu if more than 25 pages are provided
+        # this will show the previous 12 and next 13 pages in the select menu
+        # based on the currently displayed page. Once you reach close to the max
+        # pages it will display the last 25 pages.
+        if len(self.select_options) > 25:
+            minus_diff = None
+            plus_diff = 25
+            if 12 < self.current_page < len(self.select_options) - 25:
+                minus_diff = self.current_page - 12
+                plus_diff = self.current_page + 13
+            elif self.current_page >= len(self.select_options) - 25:
+                minus_diff = len(self.select_options) - 25
+                plus_diff = None
+            options = self.select_options[minus_diff:plus_diff]
+        else:
+            options = self.select_options[:25]
+        return _SelectMenu(options)
+
+    async def start(self, ctx: Context, *, ephemeral: bool = False):
+        """
+        Used to start the menu displaying the first page requested.
+
+        Parameters
+        ----------
+            ctx: `commands.Context`
+                The context to start the menu in.
+            ephemeral: `bool`
+                Send the message ephemerally. This only works
+                if the context is from a slash command interaction.
+        """
+        self.author = ctx.author
+        self.ctx = ctx
+        kwargs = await self.get_page(self.current_page)
+        self.message = await ctx.send(**kwargs, ephemeral=ephemeral)
```

Comment:
I'm not sure if you plan on making this PR depend on a higher version of discord.py, but for the version of discord.py that Red currently installs this raises a `TypeError` for the unexpected `ephemeral=` kwarg.

Reply:
I expect that by the time this PR gets fully reviewed and merged we will have bumped our d.py version to one that includes this kwarg on ctx. I added this as a sort of future-proof in the event cog creators want to use the menu as an ephemeral response from a slash command.